src/wifi/model/periodic-dmg-wifi-scheduler.h

```diff
@@ -0,0 +1,113 @@
+  /**
+   * Implement the policy that accepts or rejects a new ADDTS request.
+   * \param sourceAid The AID of the requesting STA.
+   * \param dmgTspec The DMG Tspec element of the ADDTS request.
+   * \param info The DMG Allocation Info element of the request.
+   * \return The Status Code to be included in the ADDTS response.
+   */
+  virtual StatusCode AddNewAllocation (uint8_t sourceAid, const DmgTspecElement &dmgTspec, const DmgAllocationInfo &info);
+  /**
+   * Implement the policy that accepts or rejects a modification request.
+   * The current version supports only requests of reduction of the duration of the allocations.
+   * \param sourceAid The AID of the requesting STA.
+   * \param dmgTspec The DMG Tspec element of the ADDTS request.
+   * \param info The DMG Allocation Info element of the request.
+   * \return The StatusCode to be included in the ADDTS response.
+   */
+  virtual StatusCode ModifyExistingAllocation (uint8_t sourceAid, const DmgTspecElement &dmgTspec, const DmgAllocationInfo &info);
+  /**
+   * Adjust the existing allocations when an allocation is removed or modified.
+   * \param iter The iterator pointing to the next element in the addtsAllocationList.
+   * \param duration The duration of the time to manage.
+   * \param isToAdd Whether the duration is to be added or subtracted.
+   */
+  virtual void AdjustExistingAllocations (AllocationFieldListI iter, uint32_t duration, bool isToAdd);
+  /**
+   * Update start time and remaining DTI time for the next request to be evaluated.
+   */
+  virtual void UpdateStartAndRemainingTime (void);
+  /**
+   * Add broadcast CBAP allocations in the DTI.
+   */
+  virtual void AddBroadcastCbapAllocations (void);
+
+private:
+  /**
+   * Verify how many blocks (in our case, one SP corresponds to one block) we can guarantee to a periodic request.
+   * \param allocDuration the duration associated to the SPs.
+   * \param spInterval time between two consecutive periodic SPs.
+   * \return number of blocks that can be allocated in the DTI
+   */
+  uint8_t GetAvailableBlocks (uint32_t allocDuration, uint32_t spInterval);
+  /**
+   * Update the list of available time slots in the DTI. 
+   * The current version supports only the reduction of pre-existing allocations.
+   * \param startAlloc start time of the allocation that has to be excluded from the available time.
+   * \param endAlloc end time of the allocation that has to be excluded from the available time.
+   * \param difference if not equal to 0, it represents how much an allocation has been reduced.
+   */
+  void UpdateAvailableSlots (uint32_t startAlloc, uint32_t endAlloc, uint32_t difference = 0);
+
+  // std::pair<uint32_t, uint32_t> is a struct used to store the start and 
+  // end time (first and second member of the pair, respectively) of the available time chunks
+  std::vector<std::pair<uint32_t, uint32_t> > m_availableSlots;       //!< List of available time chunks in the DTI.
```

Comment:
From this point of view, my solution is the simplest possible. No particular optimization techniques have been implemented :) You raised a good point, I'll try and see what can be done to optimize as you suggested

Reply:
if you want to get over with this quickly (which i would totally agreed on), just using `std::list` instead of `std::vector` would suffice. It's the brute force way, but at least you use the proper data structure for that :)